Adding "driver" and "dangling" filter options to "docker volume ls"

dani-docker opened the pull request:
`volume ls` in swarm classic only handles filters with options "node", "name" and "label".
This PR adds support for "driver" and "dangling"

Adding "driver" to the filter option is pretty straightforward.
The "dangling" filter is a bit more involved as the "volume" struct has no "dangling" property.

2 options to support "dangling":
1- add an extra call to the engine to get the dangling volumes
or
2- Iterate through cached "containers" and build a map of volumes that are still referenced by the container, and finally use the delta to delta to determine dangling volumes.

After chatting with @nishanttotla , 1) seems a bit taxing so I decided to go with option 2)
_Note: I am not caching the map for referenced volumes, so we don't have to maintain it, besides, the calls to filter based on "dangling" should be fast and should not be that frequent._ 

**_Repro Steps for `driver`:_**

1- Have at least 2 volumes using 2 separate drivers. 
for ex:
docker volume ls

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|[vieux/sshfs:latest](https://github.com/vieux/docker-volume-sshfs)|sshvolume|
|local                |     testVolume           |

2- Run 
`docker volume ls --filter=driver=local`

3- 
_Expect Results_

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|local                |     testVolume           |

_Actual Results (all volumes returned)_

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|vieux/sshfs:latest|sshvolume|
|local                |     testVolume           |

_**Repro Steps for `dangling`:**_

1- Have at least 2 volumes with one volume that is dangling (not used).
(sshvolume in this example is dangling)
docker volume ls

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|vieux/sshfs:latest|sshvolume|
|local                |     testVolume           |

2- Run
`docker volume ls --filter=dangling=true`

3- 
_Expect Results_

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|vieux/sshfs:latest|sshvolume|

_Actual Results (all volumes returned)_

|DRIVER           |     VOLUME NAME    |
|--------------|---------------------|
|vieux/sshfs:latest|sshvolume|
|local                |     testVolume           |




Signed-off-by: Dani Louca <dani.louca@docker.com>

wsong commented on api/handlers.go:
```diff
@@ -359,6 +383,18 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
 
+		found = false
+		for _, driver := range drivers {
+			if strings.Contains(volume.Driver, driver) {
```
The daemon code uses `filter.ExactMatch` instead of a substring match here:

https://github.com/moby/moby/blob/a3efe9722f34af5cf4443fe3a5c4e4e3e0457b54/daemon/list.go#L636

dani-docker replied:
makes sense, I will correct that

dani-docker replied:
...done

wsong commented on api/handlers.go:
```diff
@@ -341,6 +341,30 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
+			for _, container := range c.cluster.Containers() {
+				for _, mount := range container.Mounts {
+					if mount.Type == "volume" {
+						if mount.Driver == "local" {
```
Does this mean that the `dangling` filter will not work for non-local volumes?

dani-docker replied:
no, it still works for all, local and non-local. 
The condition on line 353 is there to honor the "existing" naming convention used around line 417 for local volumes

https://github.com/dani-docker/swarm/blob/9ae1c6e16226bceae46afb551a3ed3287c8d6176/api/handlers.go#L405-L418

wsong commented on api/handlers.go:
```diff
@@ -359,6 +383,18 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
 
+		found = false
+		for _, driver := range drivers {
+			if volume.Driver == driver {
```
Is there any reason you can't use `filters.ExactMatch`, just like the daemon code?

dani-docker replied:
no reason, I was just following the existing pattern in `name` and `node` matching, but yes, using ExactMatch is much cleaner. 
shall I change the name/node filter matching to mirror the daemon? ie using `filters.Match`?
or "if it ain't broke don't fix it"

wsong replied:
Yeah, let's leave the name/node filtering alone. Let's keep this PR focused.

dani-docker replied:
sounds good! 
PR is updated with ExactMatch

allencloud commented on api/handlers.go:
```diff
@@ -341,6 +341,29 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
+	danglingOnly := false
+	refVolumes := make(map[string]bool)
+
+	if filters.Include("dangling") {
```
Actually I do not like so much ident. Could rewrite this to use return fast to make it much more readable?

if

allencloud commented on api/handlers.go:
```diff
@@ -359,7 +382,13 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
 		}
 
-		if filters.Include("label") {
+                if filters.Include("driver") {
```
please use `gofmt` to format these code.

thaJeztah commented on api/handlers.go:
```diff
@@ -359,6 +382,12 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
 			continue
 		}
 
+		if filters.Include("driver") {
```
Instead of filtering in swarm, wouldn't it be possible to just have the engines do this filtering? The engine API already provides these filters, so should be able to return just what we're looking for?

(I'm not to familiar with the Swarm code-base, but always assumed that's what it did, but I now see all engine's functionality is replicated here)

dani-docker replied:
The filtering is done on locally cached volumes, making calls to all engines in the cluster to filter out accordingly defeats the swarm classic design. Note, this is not specific to Volumes, same applies to Images, Containers, Networks etc...

wsong replied:
Swarm classic is designed to always read from the cache (and do its own filtering) on read-only calls. For performance reasons, we cannot do engine calls during read-only operations.

thaJeztah commented on api/handlers.go:
```diff
@@ -341,6 +341,29 @@ func getVolumes(c *context, w http.ResponseWriter, r *http.Request) {
+
+	if filters.Include("dangling") {
+		if filters.ExactMatch("dangling", "true") {
+			danglingOnly = true
```
Note that the `dangling` filter also supports `dangling=false`, which returns all volumes that are _not_ dangling (that case could be handled by passing the filter to the engine, and collect all non-dangling volumes)

dani-docker replied:
Right, calling the engine directly was one simple option (see PR description), but api calls to all engines in the cluster can be taxing, hence we decided to build the list of used volumes from locally cached containers.

thaJeztah replied:
Won't that easily lead to incorrect data (i.e. container may now use the volume, or vice-versa); or not a problem?

dani-docker replied:
good catch @thaJeztah! I will address the above

dani-docker commented:
had a chat with @thaJeztah and 2 valid use cases were identified:

1. dangling=false has not been handled in my PR, I was under the impression that `false` means don't do any filtering, but it actually means to return used volumes only . Verified and confirmed the behavior directly on the engine

2. Existing bug that was surfaced with the `dangling` fix. Currently, we only append the engine name to `local` volumes, this name uniqueness should be applied to all type of drivers otherwise we can see false/positive results.

I will update the PR shortly.

dani-docker commented:
While waiting on #2832 
- I changed the code to address the below internally without changing the output

> Existing bug that was surfaced with the dangling fix. Currently, we only append the engine name to local volumes, this name uniqueness should be applied to all type of drivers otherwise we can see false/positive results.

- added 2 helper methods along with their unit tests 

if #2832 is approved, I will update this PR with cleaner/simpler code
 ping @wsong @nishanttotla

wsong commented on api/handlers.go:
```diff
@@ -330,6 +330,39 @@ func getVolume(c *context, w http.ResponseWriter, r *http.Request) {
 	httpError(w, fmt.Sprintf("No such volume: %s", name), http.StatusNotFound)
 }
 
+func GetUsedVolumes(containers []*cluster.Container) map[string]bool {
```
This function doesn't need to be public (i.e. it can be named `getUsedVolumes` instead of `GetUsedVolumes`.

dani-docker replied:
right, the unit test is in the same package, no need for public.
I will fix that

wsong commented on api/handlers.go:
```diff
@@ -330,6 +330,39 @@ func getVolume(c *context, w http.ResponseWriter, r *http.Request) {
+	return usedVolumes
+}
+
+func GetFilteredVolumes(containers []*cluster.Container, volumes []*apitypes.Volume, dangling bool) []*apitypes.Volume {
```
As above, make this function private.

dani-docker replied:
right, the unit test is in the same package, no need for public.
I will fix that

wsong commented on api/handlers.go:
```diff
@@ -330,6 +330,39 @@ func getVolume(c *context, w http.ResponseWriter, r *http.Request) {
+	// volume name allowed char [a-zA-Z0-9][a-zA-Z0-9_.-]
+	// splitting on / is safe to keep the original output intact
+	stripEngine := func(vol *apitypes.Volume) *apitypes.Volume {
+		if vol.Driver != "local" {
```
Instead of special-casing local volumes, can you just do `parts = strings.Split(vol.Name)` and use `parts[1]` if `len(parts) > 1` and `vol.Name` otherwise?

dani-docker replied:
I have to special case `local` volumes so I don't change its `engine_name/volume_name` format.

wsong replied:
In that case, add a comment.

dani-docker replied:
done

dani-docker commented:
PR updated with @wsong suggestions

wsong commented on api/handlers.go:
```diff
@@ -330,6 +330,40 @@ func getVolume(c *context, w http.ResponseWriter, r *http.Request) {
+	// volume name allowed char [a-zA-Z0-9][a-zA-Z0-9_.-]
+	// splitting on / is safe to keep the original output intact
+	stripEngine := func(vol *apitypes.Volume) *apitypes.Volume {
+		//don't change local volume names "engine_name/volume_name"
```
I meant that you should add a comment explaining why you're doing this.

dani-docker replied:
updated

dani-docker commented:
@wsong 
PR updated with the detailed comment

wsong commented on api/handlers.go:
```diff
@@ -330,6 +330,42 @@ func getVolume(c *context, w http.ResponseWriter, r *http.Request) {
+		//this naming format is temporarily extended to all volumes to provide uniqueness
+		// while identifying dangling volumes. Restoring the original naming convention back.
+		if vol.Driver != "local" {
+			vol.Name = strings.Split(vol.Name, "/")[1]
```
Would it make sense to check that `strings.Split(vol.Name, "/")` here has `len() > 1` so that we don't panic if there's a problem here?

thaJeztah replied:
Possibly SplitN as well (I think with custom drivers, a slash may be allowed in names)

dani-docker replied:
here https://github.com/dani-docker/swarm/blob/117d215985d6d06fb66115dd781e32253fe3aa79/api/handlers.go#L426 guarantees that "/" is appended when dangling filter is in use. 
But if you think this helper method will be re-used and to be on the safe, I will make the change.

dani-docker replied:
@thaJeztah 
you are correct! when I looked at the code, I only checked local volume, as you stated, custom drivers  so allow slash

dani-docker replied:
changes done for both concerns

dani-docker commented:
PR updated to address the / in plugin volume names and safeguard check for len of split.
Also updated the unit test to cover the above cases.
